### PR TITLE
[red-knot] Upgrade to the *new* *new* salsa

### PR DESCRIPTION
**Pull request by MichaReiser**
This PR upgrades Salsa and refactors our code to the [*hassle free* salsa](https://salsa.zulipchat.com/#narrow/stream/333573-salsa-3.2E0/topic/branch.3A.20.22hassle-free.22.20salsa).

Most notable changes:

* Salsa macros no longer generate clippy warnings, :partying_face: 
* `Jar` is gone. Adding `[#salsa::tracked]` is enough
* Implementing any `Db` trait requires the `[#salsa::db]` macro
* `ParallelDatabase` is gone, use `Handle` instead
* `DebugWithDb` no longer exist. Instead use the normal `Debug` implementation. Outside a query, use `db.attach(|db| dbg!(ingredient))`.

## Performance regression

I invested quiet some time in figuring out where the regression comes from but I haven't been successful so far. See [this zulip thread](https://salsa.zulipchat.com/#narrow/stream/333573-salsa-3.2E0/topic/Beautiful.20Salsa.20perf.20incremental.20checking.20perf.20regression) for the entire conversation. I don't think we should block the PR based on the regression.

There are ideas on how to mitigate the perf regression. I don't want to block this PR by it.

### CODE REVIEW

**Comment by MichaReiser** [comment]
This is mostly done. The part that's missing is to find a way to assert whether a query did run. Salsa's new model no-longer exposes the query itself. We probably need to fallback to doing string comparisons, ugh

File: crates/red_knot_python_semantic/src/semantic_index.rs [hunk 1]
```diff
@@ -47,8 +48,10 @@ pub(crate) fn semantic_index(db: &dyn Db, file: File) -> SemanticIndex<'_> {
 /// is unchanged.
 #[salsa::tracked]
 pub(crate) fn symbol_table<'db>(db: &'db dyn Db, scope: ScopeId<'db>) -> Arc<SymbolTable> {
-    let _span = tracing::trace_span!("symbol_table", ?scope).entered();
-    let index = semantic_index(db, scope.file(db));
+    let file = scope.file(db);
+    let _span =
+        tracing::trace_span!("symbol_table", scope=?scope.as_id(), file=?file.path(db)).entered();
```
Comment (carljm) [inline crates/red_knot_python_semantic/src/semantic_index.rs:53]: Snuck in some tracing improvements, too! Is this related to the Salsa change? Either way, I like it.

File: crates/red_knot_python_semantic/src/types/infer.rs [hunk 2]
```diff
@@ -2332,10 +2339,10 @@ mod tests {
 
         let events = db.take_salsa_events();
 
-        assert_function_query_was_not_run::<infer_definition_types, _, _>(
+        assert_function_query_was_not_run(
             &db,
-            |ty| &ty.function,
-            &first_public_def(&db, a, "x"),
+            infer_definition_types,
```
Comment (carljm) [inline crates/red_knot_python_semantic/src/types/infer.rs:2344]: This ended up improving the usability of the API! Nice.

File: crates/red_knot/src/main.rs [hunk 3]
```diff
@@ -125,11 +124,14 @@ pub fn main() -> anyhow::Result<()> {
...
 
+    let mut db = salsa::Handle::new(db);
     if watch {
         main_loop.watch(&mut db)?;
     } else {
         main_loop.run(&mut db);
-    }
+    };
+
+    std::mem::forget(db);
```
[the remark this answers is not in the document]
Comment (MichaReiser) [reply crates/red_knot/src/main.rs:134]: yes

File: crates/red_knot_python_semantic/src/db.rs [hunk 4]
```diff
@@ -131,26 +100,20 @@ pub(crate) mod tests {
...
-    impl salsa::ParallelDatabase for TestDb {
-        fn snapshot(&self) -> salsa::Snapshot<Self> {
-            salsa::Snapshot::new(Self {
-                storage: self.storage.snapshot(),
-                files: self.files.snapshot(),
-                system: self.system.snapshot(),
-                vendored: self.vendored.snapshot(),
-                events: self.events.clone(),
-            })
+            self.attach(|_| {
```
[the remark this answers is not in the document]
Comment (MichaReiser) [reply crates/red_knot_python_semantic/src/db.rs:112]: Yes, it uses thread local but no, Salsa also uses thread local whenever it goes from external API -> internal. That's where the perf regression comes form (at least, one of them).

File: crates/red_knot_python_semantic/src/semantic_index.rs (hunk 1 above)
Comment (MichaReiser) [reply crates/red_knot_python_semantic/src/semantic_index.rs:53]: It's because `debug` now prints the entire struct and not just the id. This was very verbose, so I had to truncate it.

File: crates/red_knot_python_semantic/src/types/infer.rs (hunk 2 above)
Comment (MichaReiser) [reply crates/red_knot_python_semantic/src/types/infer.rs:2344]: Yeah I think so. String matching is still a bit annoying but we at least have tests that tell us when the string matching stops working.
